### PR TITLE
Notification feature

### PR DESCRIPTION
**Pull request by kobakei**
## Issue
- #82 

## Overview (Required)
- This PR provides first cut of notification feature.
  - Show notification 10 minutes ago of session which user checks.
  - Cancel notification when user unchecks session.
  - By tapping the notification, app will launch.
  - If user disables notification in setting screen, the notification will not show.

## Links
-

## Screenshot
Before | After
:--: | :--:
 | dump_4582993252243571689

### CODE REVIEW

**Comment by konifar** [comment]
:eyes:

File: app/src/main/java/io/github/droidkaigi/confsched2017/receiver/NotificationReceiver.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        intent.putExtra(NotificationReceiver.KEY_SESSION_ID, sessionId);
+        intent.putExtra(NotificationReceiver.KEY_TITLE, title);
+        intent.putExtra(NotificationReceiver.KEY_TEXT, text);
+        return intent;
+    }
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        if (!DefaultPrefs.get(context).getNotificationFlag()) {
+            Log.v(TAG, "Notification is disabled.");
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/receiver/NotificationReceiver.java]: You can use `Timber.tag(TAG).v("Notification is disabled.");`.

File: app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java [hunk 2]
```diff
@@ -160,6 +166,35 @@ private String decideSessionTimeRange(Context context, Session session) {
                 DateUtil.getMinutes(displaySTime, displayETime));
     }
 
+    private PendingIntent createAlarmIntent() {
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java]: Can you create `util/AlermUtil.java` and extract these logic to the util class? 🙇

File: app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java [hunk 3]
```diff
@@ -160,6 +166,35 @@ private String decideSessionTimeRange(Context context, Session session) {
...
+                DateUtil.getHourMinute(displaySTime),
+                DateUtil.getHourMinute(displayETime),
+                room);
+        Intent intent = NotificationReceiver.createIntent(context, this.session.id, title, text);
+        return PendingIntent.getBroadcast(context, REQ_CODE_NOTIFICATION,
+                intent, PendingIntent.FLAG_UPDATE_CURRENT);
+    }
+
+    private void registerAlarm() {
+        long time = this.session.stime.getTime() - 10 * 60 * 1000;
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java]: How about using `TimeUnit`?

```java
private static final long REMIND_DURATION_MINUTES_FOR_START = TimeUnit.MINUTES.toMillis(10);
...
long time = this.session.stime.getTime() - REMIND_DURATION_MINUTES_FOR_START;
```

File: app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java [hunk 4]
```diff
@@ -139,10 +143,12 @@ public void onClickFab(@SuppressWarnings("unused") View view) {
             mySessionsRepository.delete(session)
                     .subscribe((result) -> Log.d(TAG, "Deleted my session"),
                             throwable -> Log.e(TAG, "Failed to delete my session", throwable));
+            unregisterAlarm();
         } else {
             mySessionsRepository.save(session)
                     .subscribe(() -> Log.d(TAG, "Saved my session"),
                             throwable -> Log.e(TAG, "Failed to save my session", throwable));
+            registerAlarm();
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/viewmodel/SessionDetailViewModel.java]: If the time is within 10 mins to start the session, the alerm should not be registered.
Can you check like below?

```java
public boolean shouldNotify(Session session, long remindDuration) {
    Date now = new Date();
    long diff = session.stime.getTime() - now.getTime();
    return remindDuration < diff;
}
```

File: app/src/main/res/values-ja/strings.xml [hunk 5]
```diff
@@ -39,4 +39,7 @@
     <string name="session_feedback">セッションフィードバック</string>
     <string name="info_sponsors_title">スポンサー</string>
     <string name="info_contributors_title">コントリビュータ</string>
+    <!-- Notification -->
+    <string name="notitication_title">%1$s がもうすぐ始まります。</string>
+    <string name="notification_message">%1$s ~ %2$s @ %3$s</string>
```
Comment (konifar) [inline app/src/main/res/values-ja/strings.xml]: @jfsso I'm not sure we can add the translations manually. Is it okay or should we add them from [oneskyapp](https://droidkaigi2017.oneskyapp.com)?

**Comment by konifar** [comment]
@kobakei Reviewed! Can you check the comments if you have a time?
Thanks for your quick contribution!! 👍 👍

**Comment by jfsso** [comment]
@konifar yeah, it will get overridden next time we import translations.
Please add only to the main strings.xml and translate on onesky.

File: app/src/main/res/values-ja/strings.xml (hunk 5 above)
Comment (jfsso) [inline app/src/main/res/values-ja/strings.xml]: @konifar Tried replying through email but I guess my comment got lost. Yeah, we need to add them on oneskyapp and import it later. 👍  You can leave it as is and later we add to onesky.

**Pull request closed without merging**